pandas/core/base.py

```diff
@@ -148,10 +149,18 @@ class SpecificationError(Exception):
 
 class SelectionMixin:
     """
-    mixin implementing the selection & aggregation interface on a group-like
+    Mixin implementing the selection & aggregation interface on a group-like
     object sub-classes need to define: obj, exclusions
+
+    obj : FrameOrSeries
+        Target object for the selection and aggregation.
+    exclusions : set, optional
+        Columns to exclude.
     """
 
+    obj: Any
```

Comment:
`obj` here is `Any` and `FrameOrSeries` in docstring.

Reply:
Yes, this looks odd.
Initially I typed ``obj`` as ``FrameOrSeries``, but then mypy would raise errors of this kind.

```
pandas\core\window\rolling.py:228: error: FrameOrSeries? has no attribute "ndim"  [attr-defined]
pandas\core\window\rolling.py:229: error: FrameOrSeries? has no attribute "__iter__" (not iterable)  [attr-defined]
pandas\core\window\rolling.py:236: error: FrameOrSeries? has no attribute "__iter__" (not iterable)  [attr-defined]
pandas\core\window\rolling.py:244: error: FrameOrSeries? has no attribute "_dir_additions"  [attr-defined]
pandas\core\window\rolling.py:324: error: FrameOrSeries? has no attribute "index"  [attr-defined]
```

The reason is that ``DataFrame`` and ``Series`` interface is similar, but different.
Some attributes are defined in one, but not defined in another, or the methods have different signatures.

Reply:
FrameOrSeries is a typevar, so to use it on a class/instance variable requires to make the SelectionMixin class generic. The mypy error suggests this.

```
pandas/core/base.py:161: error: Type variable "pandas._typing.FrameOrSeries" is unbound  [valid-type]
pandas/core/base.py:161: note: (Hint: Use "Generic[FrameOrSeries]" or "Protocol[FrameOrSeries]" base class to bind "FrameOrSeries" inside a class)
pandas/core/base.py:161: note: (Hint: Use "FrameOrSeries" in function signature to bind "FrameOrSeries" inside a function)
```

if having self.obj not maintain it's type for the duration of the instance life is ok, then use FrameOrSeriesUnion instead.

mypy errors such as `pandas/core/window/rolling.py:108: error: Incompatible types in assignment (expression has type "FrameOrSeries", variable has type "Union[DataFrame, Series]")  [assignment]` if doing this are where FrameOrSeries aliases are used elsewhere that also would need to change to FrameOrSeriesUnion.

Reply:
There are 34 mypy errors of this kind (see details below) when setting to ``FrameOrSeriesUnion``.
Would it be a good idea for the follow up PR to fix them?


<details>
pandas\core\window\rolling.py:108: error: Incompatible types in assignment (expression has type "FrameOrSeries", variable has type "Union[DataFrame, Series]")  [assignment]
pandas\core\groupby\groupby.py:571: error: Incompatible types in assignment (expression has type "FrameOrSeries", variable has type "Union[DataFrame, Series]")  [assignment]
pandas\core\groupby\groupby.py:872: error: Argument 1 to "get_iterator" of "BaseGrouper" has incompatible type "Union[DataFrame, Series]"; expected "FrameOrSeries"  [arg-type]
pandas\core\groupby\generic.py:233: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:358: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:440: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:467: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:501: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:567: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:567: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "Series")  [return-value]
pandas\core\groupby\generic.py:668: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:669: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "Series")  [return-value]
pandas\core\groupby\generic.py:777: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:809: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:828: error: Unexpected keyword argument "name" for "DataFrame"  [call-arg]
pandas\core\frame.py:511: note: "DataFrame" defined here
pandas\core\groupby\generic.py:834: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "Series")  [return-value]
pandas\core\groupby\generic.py:947: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1171: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1171: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "DataFrame")  [return-value]
pandas\core\groupby\generic.py:1278: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1289: error: Argument 1 to "_insert_inaxis_grouper_inplace" of "DataFrameGroupBy" has incompatible type "Union[DataFrame, Series]"; expected "DataFrame"  [arg-type]
pandas\core\groupby\generic.py:1325: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1353: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1398: error: Item "type" of "Union[Type[DataFrame], Type[Series]]" has no attribute "_from_arrays"  [union-attr]
pandas\core\groupby\generic.py:1462: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1462: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "DataFrame")  [return-value]
pandas\core\groupby\generic.py:1597: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1601: error: Unexpected keyword argument "columns" for "Series"  [call-arg]
pandas\core\series.py:218: note: "Series" defined here
pandas\core\groupby\generic.py:1601: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "DataFrame")  [return-value]
pandas\core\groupby\generic.py:1648: error: Argument 1 to "_insert_inaxis_grouper_inplace" of "DataFrameGroupBy" has incompatible type "Union[DataFrame, Series]"; expected "DataFrame"  [arg-type]
pandas\core\groupby\generic.py:1656: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "DataFrame")  [return-value]
pandas\core\groupby\generic.py:1686: error: Incompatible return value type (got "Union[DataFrame, Series]", expected "DataFrame")  [return-value]
pandas\core\groupby\generic.py:1694: error: Argument 1 to "_insert_inaxis_grouper_inplace" of "DataFrameGroupBy" has incompatible type "Union[DataFrame, Series]"; expected "DataFrame"  [arg-type]
pandas\core\groupby\generic.py:1704: error: Incompatible return value type (got "Union[DataFrame, Any, Series]", expected "DataFrame")  [return-value]
</details>


I think I can do them here, but first need to figure out how to deal with "CI Web and docs", since it complains about the missing documentation

```
2021-01-10T11:33:04.9021813Z /home/runner/work/pandas/pandas/doc/source/reference/api/pandas.DataFrame.exclusions.rst: WARNING: document isn't included in any toctree
2021-01-10T11:33:04.9024422Z /home/runner/work/pandas/pandas/doc/source/reference/api/pandas.Series.exclusions.rst: WARNING: document isn't included in any toctree
```

Reply:
> Would it be a good idea for the follow up PR to fix them?

put a big TODO by the variable declaration otherwise this could get missed. `Any` should only be used where we want to allow dynamic typing ( and lose the benefits of static type checking)